File: src/main/java/com/example/legacyshop/service/PaymentService.java
package com.example.legacyshop.service;

import com.example.legacyshop.config.BusinessConfig;
import com.example.legacyshop.entity.OrderEntity;
import com.example.legacyshop.entity.Payment;
import com.example.legacyshop.exception.PaymentException;
import com.example.legacyshop.exception.ResourceNotFoundException;
import com.example.legacyshop.repository.OrderRepository;
import com.example.legacyshop.repository.PaymentRepository;
import org.springframework.retry.annotation.Backoff;
import org.springframework.retry.annotation.Retryable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.reactive.function.client.WebClient;
import org.springframework.web.reactive.function.client.WebClientResponseException;

import java.math.BigDecimal;
import java.time.Duration;
import java.util.Map;
import java.util.UUID;

/**
 * Payment service with retry logic and external service integration.
 * 
 * Key patterns:
 * - @Retryable for automatic retry on 5xx errors
 * - WebClient for external service calls
 * - 4xx to domain error mapping
 * - Compensation actions (void payment)
 */
@Service
@Transactional(readOnly = true)
public class PaymentService {

    private final PaymentRepository paymentRepository;
    private final OrderRepository orderRepository;
    private final WebClient webClient;
    private final BusinessConfig businessConfig;

    public PaymentService(PaymentRepository paymentRepository,
                         OrderRepository orderRepository,
                         BusinessConfig businessConfig) {
        this.paymentRepository = paymentRepository;
        this.orderRepository = orderRepository;
        this.businessConfig = businessConfig;
        this.webClient = WebClient.builder()
            .codecs(configurer -> configurer.defaultCodecs().maxInMemorySize(1024 * 1024))
            .build();
    }

    /**
     * Authorize payment with retry logic on 5xx errors.
     * Maps 4xx errors to domain exceptions.
     */
    @Transactional
    @Retryable(
        retryFor = {PaymentException.class},
        maxAttempts = 2,
        backoff = @Backoff(delay = 1000)
    )
    public Payment authorizePayment(Long orderId) {
        OrderEntity order = orderRepository.findById(orderId)
            .orElseThrow(() -> new ResourceNotFoundException("Order not found with id: " + orderId));

        // Create payment record
        Payment payment = new Payment(order, order.getTotal());
        Payment savedPayment = paymentRepository.save(payment);

        try {
            // Call external payment service
            String authorizationId = callExternalPaymentService(order.getTotal());
            
            // Update payment with success
            savedPayment.authorize(authorizationId);
            Payment result = paymentRepository.save(savedPayment);

            // Update order status
            order.markAsPaid();
            orderRepository.save(order);

            return result;

        } catch (WebClientResponseException e) {
            savedPayment.incrementRetryAttempts();
            
            if (e.getStatusCode().is4xxClientError()) {
                // 4xx errors are not retryable - map to domain error
                String errorMessage = "Payment authorization failed: " + e.getResponseBodyAsString();
                savedPayment.fail(errorMessage);
                paymentRepository.save(savedPayment);
                throw new PaymentException(errorMessage, false);
                
            } else if (e.getStatusCode().is5xxServerError()) {
                // 5xx errors are retryable
                String errorMessage = "Payment service temporarily unavailable";
                savedPayment.fail(errorMessage);
                paymentRepository.save(savedPayment);
                throw new PaymentException(errorMessage, true);
                
            } else {
                // Other errors
                String errorMessage = "Unexpected payment error: " + e.getStatusCode();
                savedPayment.fail(errorMessage);
                paymentRepository.save(savedPayment);
                throw new PaymentException(errorMessage, false);
            }
        } catch (Exception e) {
            // Network or other errors - treat as retryable
            savedPayment.incrementRetryAttempts();
            String errorMessage = "Payment authorization failed: " + e.getMessage();
            savedPayment.fail(errorMessage);
            paymentRepository.save(savedPayment);
            throw new PaymentException(errorMessage, true);
        }
    }

    /**
     * Void payment (compensation action for order cancellation).
     */
    @Transactional
    public void voidPayment(Long paymentId) {
        Payment payment = paymentRepository.findById(paymentId)
            .orElseThrow(() -> new ResourceNotFoundException("Payment not found with id: " + paymentId));

        if (payment.getStatus() != Payment.PaymentStatus.AUTHORIZED) {
            throw new IllegalStateException("Can only void authorized payments");
        }

        try {
            // Call external service to void payment
            callExternalVoidService(payment.getExternalAuthorizationId());
            
            // Update payment status
            payment.voidPayment();
            paymentRepository.save(payment);

        } catch (Exception e) {
            // Log error but don't fail the cancellation process
            // In production, this might queue for retry or alert operations
            throw new PaymentException("Failed to void payment: " + e.getMessage(), false);
        }
    }

    /**
     * Call external payment authorization service.
     * This demonstrates the integration point that could fail with 4xx/5xx.
     */
    private String callExternalPaymentService(BigDecimal amount) {
        try {
            Map<String, Object> request = Map.of(
                "amount", amount.toString(),
                "currency", "USD",
                "paymentMethod", "CARD"
            );

            Map<String, Object> response = webClient.post()
                .uri(businessConfig.getPayments().getAuthUrl())
                .bodyValue(request)
                .retrieve()
                .bodyToMono(Map.class)
                .timeout(Duration.ofSeconds(businessConfig.getPayments().getTimeoutSeconds()))
                .block();

            if (response != null && response.containsKey("authorizationId")) {
                return (String) response.get("authorizationId");
            } else {
                throw new PaymentException("Invalid response from payment service", false);
            }

        } catch (WebClientResponseException e) {
            throw e; // Re-throw to be handled by retry logic
        }
    }

    /**
     * Call external payment void service.
     */
    private void callExternalVoidService(String authorizationId) {
        Map<String, Object> request = Map.of("authorizationId", authorizationId);

        webClient.post()
            .uri(businessConfig.getPayments().getAuthUrl() + "/void")
            .bodyValue(request)
            .retrieve()
            .bodyToMono(Void.class)
            .timeout(Duration.ofSeconds(businessConfig.getPayments().getTimeoutSeconds()))
            .block();
    }

    /**
     * Get payment by order ID.
     */
    public Payment getPaymentByOrderId(Long orderId) {
        return paymentRepository.findByOrderId(orderId)
            .orElseThrow(() -> new ResourceNotFoundException("Payment not found for order: " + orderId));
    }
}

File: legacyshop-api/app/services/payment_service.py
"""
Payment service with retry logic and external service integration.
Implements retry logic matching the Java @Retryable annotation.
"""

import httpx
from tenacity import retry, stop_after_attempt, wait_fixed, retry_if_exception_type
from sqlalchemy.orm import Session
from decimal import Decimal

from app.models.models import Order, Payment, PaymentStatus
from app.config.business_config import config
from app.services.exceptions import PaymentException, ResourceNotFoundException


class PaymentService:
    """Service for payment processing with retry logic"""
    
    @retry(
        stop=stop_after_attempt(config.payment_max_attempts),
        wait=wait_fixed(config.payment_backoff_seconds),
        retry=retry_if_exception_type(PaymentException),
        reraise=True
    )
    async def authorize_payment(self, order_id: int, db: Session) -> Payment:
        """
        Authorize payment with retry logic on 5xx errors.
        Maps 4xx errors to domain exceptions.
        """
        order = db.query(Order).filter(Order.id == order_id).first()
        if not order:
            raise ResourceNotFoundException(f"Order not found with id: {order_id}")
        
        existing_payment = db.query(Payment).filter(Payment.order_id == order.id).first()
        if existing_payment:
            if existing_payment.status == PaymentStatus.AUTHORIZED:
                return existing_payment
            elif existing_payment.status == PaymentStatus.PENDING:
                payment = existing_payment
            else:
                payment = Payment(
                    order_id=order.id,
                    amount=order.total,
                    status=PaymentStatus.PENDING
                )
                db.add(payment)
                db.commit()
                db.refresh(payment)
        else:
            payment = Payment(
                order_id=order.id,
                amount=order.total,
                status=PaymentStatus.PENDING
            )
            db.add(payment)
            db.commit()
            db.refresh(payment)
        
        try:
            authorization_id = await self._call_external_payment_service(order.total)
            
            payment.authorize(authorization_id)
            db.commit()
            
            order.mark_as_paid()
            db.commit()
            
            return payment
            
        except httpx.HTTPStatusError as e:
            payment.increment_retry_attempts()
            
            if 400 <= e.response.status_code < 500:
                error_message = f"Payment authorization failed: {e.response.text}"
                payment.fail(error_message)
                db.commit()
                raise PaymentException(error_message, retryable=False)
            
            elif 500 <= e.response.status_code < 600:
                error_message = "Payment service temporarily unavailable"
                payment.fail(error_message)
                db.commit()
                raise PaymentException(error_message, retryable=True)
            
            else:
                error_message = f"Unexpected payment error: {e.response.status_code}"
                payment.fail(error_message)
                db.commit()
                raise PaymentException(error_message, retryable=False)
                
        except Exception as e:
            payment.increment_retry_attempts()
            error_message = f"Payment authorization failed: {str(e)}"
            payment.fail(error_message)
            db.commit()
            raise PaymentException(error_message, retryable=True)
    
    async def _call_external_payment_service(self, amount: float) -> str:
        """
        Call external payment authorization service.
        This demonstrates the integration point that could fail with 4xx/5xx.
        """
        async with httpx.AsyncClient(timeout=config.payment_timeout_seconds) as client:
            request_data = {
                "amount": str(amount),
                "currency": "USD",
                "paymentMethod": "CARD"
            }
            
            response = await client.post(
                config.payment_auth_url,
                json=request_data
            )
            response.raise_for_status()
            
            data = response.json()
            if "authorizationId" in data:
                return data["authorizationId"]
            else:
                raise PaymentException("Invalid response from payment service", retryable=False)
    
    def void_payment(self, payment_id: int, db: Session):
        """
        Void payment (compensation action for order cancellation).
        """
        payment = db.query(Payment).filter(Payment.id == payment_id).first()
        if not payment:
            raise ResourceNotFoundException(f"Payment not found with id: {payment_id}")
        
        if payment.status != PaymentStatus.AUTHORIZED:
            raise ValueError("Can only void authorized payments")
        
        try:
            payment.void()
            db.commit()
        except Exception as e:
            raise PaymentException(f"Failed to void payment: {str(e)}", retryable=False)


payment_service = PaymentService()
